### ISSUE TITLE
Add support for Windows 11 "Snap Layout" to the custom title bar (WindowChrome)

### ISSUE THREAD

**Issue by d2phap** [dotnet/wpf#4825]
As Windows 11 introduce **Snap Layout** when we hover on Restore/Maximize caption button, I want to have this feature on my custom title bar using `WindowChrome`.

The current `SystemCommands` class does not have it:
```cs
public static class SystemCommands
{
  public static void CloseWindow(Window window);
  public static void MaximizeWindow(Window window);
  public static void MinimizeWindow(Window window);
  public static void RestoreWindow(Window window);
  public static void ShowSystemMenu(Window window, Point screenLocation);
}
```

### Custom Title bar
![image](https://user-images.githubusercontent.com/3154213/125071292-aed40900-e0eb-11eb-9953-1305c79fa393.png)

### Default Title bar
![image](https://user-images.githubusercontent.com/3154213/125071422-d4611280-e0eb-11eb-8300-3ea3dc36e947.png)

**Comment by lindexi** [comment]
Ref https://github.com/dotnet/wpf/issues/4749

**Comment by batzen** [comment]
Does anyone know if there is an API to trigger the new snap layout "menu"?

**Comment by miloush** [comment]
It is not a system command.

If you have a custom maximize button, it is your job to tell that to the OS by responding HTMAXBUTTON to WM_NCIHITTEST. That will give you the snap options on hover.

I think there is an opportunity for API improvement to simplify this task through an attached `WindowChrome` property

**Comment by mdtauk** [comment]
> It is not a system command.
> 
> If you have a custom maximize button, it is your job to tell that to the OS by responding `HTMAXBUTTON` to `WM_NCIHITTEST`. That will give you the snap options on hover.
> 
> I think there is an opportunity for API improvement to simplify this task through an attached `WindowChrome` property

This would be helpful for Windows Terminal, as well as other WinUI 3 apps.

**Comment by batzen** [comment]
@miloush While returning `HT.MAXBUTTON` works and shows the menu it causes various unwanted side effects.
- MouseOver does not work anymore (highlight)
- Clicking a button which returned some window button `HT` causes the system button to be rendered for a short time

![WPF_HT_Button](https://user-images.githubusercontent.com/1971030/125199519-d69bac00-e266-11eb-8d25-dff7db3035c5.gif)

**Comment by miloush** [comment]
@batzen I guess you can still handle the mouse over, however yeah I will give you that the system-rendered button on click is sad; feels like a bug in the shell.

**Comment by Splitwirez** [comment]
@miloush The events aren't even raised when the caption button `HT` values are returned. The WPF stuff just acts like the cursor isn't even over the window.

**Comment by billhenn** [comment]
We also have been digging into this issue of trying to show the Snap Layout Menu in a WPF app that uses our custom WindowChrome, where we render all custom title bar buttons.  We've run into the same issues when returning HTMAXBUTTON from WM_NCHITTEST as described above, where WPF no longer receives any pointer input (due to HTCLIENT not being returned), and clicking the our custom button renders a classic Win32 maximize button.

**Comment by batzen** [comment]
FYI: The WindowChromeBehavior from ControlzEx will provide a solution for this problem during Hacktoberfest.

**Comment by mevey** [comment]
@batzen Are you able to share with us a copy of the test app so we can get to the bottom of this?

**Comment by batzen** [comment]
I can share a link to the branch in ControlzEx as soon as i pushed the changes.
Feel free to ping me if i didn't do so till monday.

**Comment by batzen** [comment]
@mevey You can grab the code from https://github.com/ControlzEx/ControlzEx/tree/features/Win11Support i haven't yet implemented support for rounded corners, but the snap menu on the maximize button works without all the negative side effects everyone observed.

**Comment by mevey** [comment]
Thank you @batzen, this is extremely helpful as we try to understand root cause.

**Comment by billhenn** [comment]
We also have worked the past week on Windows 11 features in our own WindowChrome and have managed to get the snap layout menu showing without the two negative side effects we found in our first attempt.  Our changes can be summarized as:

1) Handle WM_NCHITTEST and return HTMAXBUTTON when over our custom WPF Button that represents the Maximize button in our window chrome's title bar.  This gets the system to show the snap layout menu when the mouse is over our Maximize button.

2) Mark the WM_NCLBUTTONDOWN message as handled when over HTMAXBUTTON per number 1 above.  This prevents the classic 3D Win32 button from showing when clicking our WPF Maximize button.

3) Handling various WM_NC* mouse messages and using SendMessage to send related messages to the Window to mimic "client area" mouse messages with translated POINT values so that WPF can receive them.  This ensures the pressed state can be reflected on the button.  But it doesn't help with the hover state since Win32 still thinks we are in a non-client area so WM_MOUSELEAVE gets fired by the system immediately any time we send WM_MOUSEMOVE messages ourselves, thereby making WPF's IsMouseOver fail.

While the above sounds relatively simple, it really isn't and ended up being a lot more work than we'd hoped.  We've had to hack things in to handle the edge cases like proper hover state display on the button when the menu is open.  It would be nice if the system could notify the window somehow when the snap layout menu was opening/closing so we could alter our hover state on our Maximize button.  Right now it seems no Windows Win32 messages are sent to the window once the snap layout menu opens.  It's difficult to know if the mouse then moves over that menu or not.

**Comment by batzen** [comment]
@billhenn Point 3 is the reason why I created https://github.com/ControlzEx/ControlzEx/blob/features/Win11Support/src/ControlzEx/Behaviors/WindowChrome/NonClientControlManager.cs 
It's simply not possible, judging from all my experiments, to return the proper hittest result and get client area mouse messages at the same time as the area is treated as non client after the hittest result.

**Comment by billhenn** [comment]
> It's simply not possible, judging from all my experiments, to return the proper hittest result and get client area mouse messages at the same time as the area is treated as non client after the hittest result.

Yes, that is the one of the most difficult things we've had to attempt to work around.  It would be nice if there was a way to manually request the snap layout menu show so we could avoid the entire hacky non-client piece of things.  If our WPF Button could detect a hover scenario on itself, then we could simply call some Windows API (which probably doesn't currently exist) to request the snap layout menu at a certain screen coordinate and be done with it.

**Comment by mevey** [comment]
Hey folks, Just letting you know that we are still working on this issue internally and will have a response for you soon. We appreciate your patience.

**Comment by ghost1372** [comment]
Hi @batzen Your example was very complex (at least for me) I used another code. I was able to show SnapLayout But I see the problems you mentioned earlier
![00](https://user-images.githubusercontent.com/9213496/139677361-589fb42f-b28a-404a-9c5f-2842e86dedaa.gif)
 
You seem to have solved these problems, Can you help me solve these problems?

```
int x = lparam.ToInt32() & 0xffff;
int y = lparam.ToInt32() >> 16;
var rect = new Rect(_ButtonMax.PointToScreen(new Point()), new Size(_ButtonMax.Width, _ButtonMax.Height));
if (rect.Contains(new Point(x, y)))
{
    handled = true;
}
return new IntPtr(HTMAXBUTTON);
```

**Comment by sachinkumarrajak08** [comment]
Hi, @ghost1372  . i have checked your code in my  c# wpf application ,i am getting compile time error .
can you tell me or provide any sample for  how to use your code in c# wpf for button control .

**Comment by ghost1372** [comment]
> Hi, @ghost1372 . i have checked your code in my c# wpf application ,i am getting compile time error . can you tell me or provide any sample for how to use your code in c# wpf for button control .

you should write this code in wndProc see [here](https://github.com/ghost1372/HandyControls/blob/99b7cd169c0dbed3b8ff69311fc49e49575fd5b1/src/Shared/HandyControl_Shared/Controls/Window/Window.cs#L411-L418)

**Comment by sachinkumarrajak08** [comment]
Hi, @ghost1372 , i have checked your given link in my apps, my apps  does not return  IntPtr(HTMAXBUTTON) results.

**Comment by ghost1372** [comment]
> Hi, @ghost1372 , i have checked your given link in my apps, my apps does not return IntPtr(HTMAXBUTTON) results.

You have to pay attention to the **HwndSourceHook** function and how to register it in         
`protected override void OnSourceInitialized(EventArgs e)
`

Follow the tutorial in the second [link](https://qiita.com/dhq_boiler/items/ec5cca9a9adb37feacdb)

**Comment by sachinkumarrajak08** [comment]
Hi, @ghost1372 , i have added this below code for solving extra max button showing issue. it has fixed that issue but i am unbale to resize my apps windows width  and Hight size using my mouse.
if (msg == WM_NCLBUTTONDOWN)
            {
                 handled = true; ;
             }

**Comment by ghost1372** [comment]
@sachinkumarrajak08  yes i have this issue too, also you mouse hover not working for max button

**Comment by Pomianowski** [comment]
To trigger a click in a non-client area we can do it this way

```c#
case Input.WM_NCLBUTTONDOWN:
  IInvokeProvider invokeProv = new ButtonAutomationPeer(_myButtonControl)
    .GetPattern(PatternInterface.Invoke) as IInvokeProvider;
  invokeProv?.Invoke();

  handled = true;

  break;
```

**Comment by ghost1372** [comment]
Hi @Pomianowski Thank you your code worked fine, But there is one problem
One click on titlebar maximizes Window
![01](https://user-images.githubusercontent.com/9213496/143905825-d98546b9-06c8-4a37-afca-934d0a973865.gif)


and Do you have any solution for button hover?
![03](https://user-images.githubusercontent.com/9213496/143905944-640df780-9a15-4531-9b16-538821462cbc.gif)

**Comment by Pomianowski** [comment]
Hi @ghost1372, initially I did something like that:
https://github.com/lepoco/wpfui/blob/main/WPFUI/Common/SnapLayout.cs

**Comment by ghost1372** [comment]
@Pomianowski  thanks
but I found another way and everything works fine😁
If anyone is interested, you can see the codes here
https://github.com/ghost1372/HandyControls/commit/41fce1df04b45ab9a7a8ebad33f3810a89a1ad13

**Comment by gurpreet-wpf** [comment]
@d2phap It is available in latest build, you can test this. I hope it is alright to go ahead and close this.

**Comment by batzen** [comment]
@gurpreet-wpf Which build are you referring to?

**Comment by kalin-todorov** [comment]
@gurpreet-wpf Can you please share more info about the exact change and how to test it?

**Comment by gurpreet-wpf** [comment]
Below are the details where you can test this:
Windows version : Windows 11 Pro 21H2
.Net version : 6.0

**Comment by kalin-todorov** [comment]
Hey @gurpreet-wpf thanks for those details. However it is not clear how this should be implemented and what is the actual change? Do you have working example of Window with custom title bar supporting Snap Layout menu?

**Comment by gurpreet-wpf** [comment]
Hi @kalin-todorov , the default implementation of this feature got fixed by Windows team. We need to work on Custom Chrome bar to have this ability.

**Comment by kalin-todorov** [comment]
Hi @gurpreet-wpf, so is there any new API that will allow us to show manually the menu?

**Comment by kalin-todorov** [comment]
Hi again @gurpreet-wpf any updates on this item?

**Comment by pchaurasia14** [comment]
@kalin-todorov  - Apologies for the delayed response. 
Unfortunately, we don't have the snap UI support for custom titlebar, yet. 

However, we're planning to incorporate this feature in upcoming releases.

**Comment by kalin-todorov** [comment]
@pchaurasia14 thanks for the reply! However, I'm more interested in the Windows API that will allow custom title bar to be implemented in a correct manner. So, is there any new Windows API that can be used to open up the snap UI manually?

**Comment by sungaila** [comment]
> So, is there any new Windows API that can be used to open up the snap UI manually?

I am also interested in this Win32/WinRT API. Visual Studio is built with WPF, has custom titlebar buttons AND the Snap Layout panel is working just fine. How did they do it?

**Comment by batzen** [comment]
The requirement to get the snap menu is to return the proper hittest result when Windows asks your app.
As far as I know there is no dedicated API.

**Comment by GSonofNun** [comment]
I got this half working using examples people have provided. If the window is maximized, the snap layout popup appears when hovering over the "Restore" button. But if the window is not maximized, it doesn't every show, even though I am still returning HTMAXBUTTON. Has anyone else seen this issue?

**Comment by manfromarce** [comment]
I've experimented that if you have a custom chrome but WindowStyle is not None the snap popup is only displayed if you return HTMAXBUTTON where the system button would be. That means, if the title bar buttons are smaller/larger/taller than the system ones it will only work on part of the maximize button. It works properly if WindowStyle=None, but that requires more code for fixing maximization, flicker and other stuff.

**Comment by GSonofNun** [comment]
@manfromarce you were right. Changing my WindowStyle to None made it start working when the window is not maximized. But unfortunately I don't want to use WindowStyle=None because of what you lose when you do. Hopefully this is a bug that they can fix. (@gurpreet-wpf)

**Comment by Gilfoylex** [comment]
https://github.com/Gilfoylex/WPFUiBase Here's my solution, referenced https://github.com/grassator/win32-window-custom-titlebar and https://github.com/lepoco/wpfui

**Comment by NotYoojun** [comment]
Hi! We're developing a library that enables us to use modern styles and components in WPF applications.
It contains a whole set to modernize your application. And it's still active.
The modern window in this lib has all snap supports. See more at:

https://github.com/InkoreStudios/UI.WPF.Modern

Thanks!

**Comment by NotYoojun** [comment]
> @Gilfoylex Good, looking forward to your https://github.com/Gilfoylex/WPFUiBase

嘿嘿，到处都是你 😀

**Comment by damien-c-d** [comment]
It would be nice to have this be possible. The default title bar in WPF gives off big Windows 7 vibes and seems to have been completely ignored by Microsoft.

**Comment by batzen** [comment]
@damien-c-d The default title bar in WPF is the default Windows title bar and supports snap layout. But when using a custom window chrome there is no snap layout and a plethora of other issues caused by it.
So i am not sure what you are talking about, when referring to "default title bar".
Could you clarify that?

**Comment by damien-c-d** [comment]
> @damien-c-d The default title bar in WPF is the default Windows title bar and supports snap layout. But when using a custom window chrome there is no snap layout and a plethora of other issues caused by it.
> So i am not sure what you are talking about, when referring to "default title bar".
> Could you clarify that?

The default title bar is rubbish and the offers very little options in the way of customisation and theme support. That's why an overwhelming majority have customised the window chrome.

You're completely correct, using a custom window chrome does have those problems. Seems like a design issue to me. So really that's the only out of the box options we have in WPF. Pretty poor in my opinion.

**Comment by Greedquest** [comment]
> Hey folks, Just letting you know that we are still working on this issue internally and will have a response for you soon. We appreciate your patience.

@mevey Any further update on this?

**Comment by sungaila** [comment]
> The default title bar is rubbish and the offers very little options in the way of customisation and theme support. That's why an overwhelming majority of UI libraries have customised the window chrome.

I'd rather stick with the default titlebar if WPF would take care of most boilerplate code. E.g. query the Windows theme settings and to set `DWMWA_USE_IMMERSIVE_DARK_MODE` (for the native dark themed titlebar) accordingly.

**Comment by mortenbrudvik** [comment]
[ControlzEx WindowChromeBehavior](https://github.com/ControlzEx/ControlzEx) works like a charm, fully supporting snap layout. I have also created a fully custom window titlebar example based on ControlzEx. [code example](https://github.com/mortenbrudvik/custom-wpf-window-titlebar-with-snap-layout)

**Comment by Yari27** [comment]
After implementing Snap Layout, styles for the button do not work.
How do I force a button hover (color) style to be used (for used Triggers in style)?

For `VisualStateManager` :
```csharp
VisualStateManager.GoToState(btnMaximizeRestore, "MouseOver", true);
VisualStateManager.GoToState(btnMaximizeRestore, "Normal", true);
```

**Comment by lindexi** [comment]
@Yari27 I'm sorry, but with limited information, I can't find the answer. Could you attach the solution folder?

**Comment by batzen** [comment]
It doesn't work because once you return hittest results like maximize the controls no longer receive mouse over events.
ControlzEx solves this by using an additional attached property for nc mouseover.

**Comment by Yari27** [comment]
@batzen thx, attached property work great.

**Comment by SlimeNull** [comment]
I create a library including this feature，process the WM_SCHITTEST message, set the IsMouseOver and IsPressed properties of the control, and correctly trigger the Button's Click event to execute its Command。

Here is the document and sample code for using it。
[Custom caption menu buttons](https://wpfsuite.elecho.dev/zh/window-options#%E8%87%AA%E5%AE%9A%E4%B9%89%E7%AA%97%E5%8F%A3%E8%8F%9C%E5%8D%95%E6%8C%89%E9%92%AE)

**Comment by augustresende** [comment]
If anyone is interested, flutter needs a way to Snap Layout work when title bar is customized too:
https://github.com/flutter/flutter/issues/156791

**Comment by icnahom** [comment]
@augustresende I believe Electron supports this feature. It'd be worth checking out how they did it.

**Comment by colhountech** [comment]
I have a version of snap layout popup working in a custom borderless window in WinForm. 

It works but it's a bit laggy and could do with some experts to review how to do it better. 

This might be a good starting point. 

You can see it in the 'features/snap-layout-improvement' branch of https://github.com:colhountechAcrylicUI/ 

Let me know if this helps.

**Comment by ignus2** [comment]
Just to chime in here because I recently did this. To correctly handle this is to respond with `HTMAXBUTTON` in `WM_NCHITTEST`. That will cause you to get the `WM_NC***` variant of the mouse messages, including `WM_NCLBUTTONDOWN`, and just make sure you prevent `DefWindowProc()` from handling `WM_NCLBUTTONDOWN` if the hittest was `HTMAXBUTTON`. You get the hittest result in the `wParam` of `WM_NCLBUTTONDOWN`. This will prevent the unwanted side-effect.

**Comment by wisd00-git** [comment]
I can't get identical behavior using the workarounds listed here. The snap menu only pops up if the mouse is over the 'real' maximize button as well. My maximize button is only partially aligned with it so about half of the time it makes the user feel like the snap feature doesn't work.

Furthermore, if I enter my button in a place that does not overlap the 'real' maximize button, stop moving my mouse, then move my cursor over to where the overlap is, the snap menu will never popup. Essentially, I have to enter my button **and** reach an overlapping point **without stopping** for the menu to show.

When the menu does popup, it is noticeable that it is centered on the 'real' maximize button as well, not my button.

Am I missing some processing of `WM_*` events or are these insurmountable limitations?

**Comment by MartinZikmund** [comment]
I'm seeing the exact same behavior as @wisd00-git . Does someone have full working sample where the the maximize button does not overlap the "system" button area, but snap layout popup still shows?

**Comment by MartinZikmund** [comment]
@wisd00-git finally figured it out! To completely avoid the "underlying" system menu, do the following:

```
var oldPtr = NativeDefinition.GetWindowLong(handle, NativeDefinition.GWL_STYLE);
var newPtr = showSystemMenu ?
    oldPtr | NativeDefinition.WS_SYSMENU :
    oldPtr & ~NativeDefinition.WS_SYSMENU;

SetWindowLong(handle, NativeDefinition.GWL_STYLE, newPtr);
```

The code is courtesy of `EleCho.WpfSuite` - thank you so much for helping me figure it out finally!
